source/mqtt/Mqtt5Listener.cpp

```diff
@@ -0,0 +1,430 @@
+                // As the constructor is private, make share would not work here. We do make_share manually.
+                Mqtt5ListenerCore *toSeat =
+                    reinterpret_cast<Mqtt5ListenerCore *>(aws_mem_acquire(allocator, sizeof(Mqtt5ListenerCore)));
+                if (!toSeat)
+                {
+                    return nullptr;
+                }
+
+                toSeat = new (toSeat) Mqtt5ListenerCore(options, client, allocator);
+                // Creation failed, make sure we release the allocated memory
+                if (!*toSeat)
+                {
+                    Crt::Delete(toSeat, allocator);
+                    return nullptr;
+                }
+
+                std::shared_ptr<Mqtt5ListenerCore> shared_listener = std::shared_ptr<Mqtt5ListenerCore>(
+                    toSeat, [allocator](Mqtt5ListenerCore *listener) { Crt::Delete(listener, allocator); });
+                shared_listener->m_selfReference = toSeat->shared_from_this();
+                return shared_listener;
+            }
+
+            Mqtt5ListenerCore::operator bool() const noexcept { return m_listener != nullptr; }
+
+            int Mqtt5ListenerCore::LastError() const noexcept { return aws_last_error(); }
+
+            void Mqtt5ListenerCore::Subscribe(Crt::String topic, Crt::Mqtt5::OnPublishReceivedHandler callback) noexcept
+            {
+                m_subscriptionMap[topic] = callback;
+            }
+
+            void Mqtt5ListenerCore::Unsubscribe(Crt::String topic) noexcept
+            {
+                if (m_subscriptionMap.find(topic) != m_subscriptionMap.end())
+                {
+                    m_subscriptionMap.erase(topic);
+                }
+            }
+
+            void Mqtt5ListenerCore::Close() noexcept
+            {
+                std::lock_guard<std::recursive_mutex> lk(m_callback_lock);
+                m_callbackFlag = CallbackFlag::IGNORE;
+                if (m_listener != nullptr)
+                {
+                    aws_mqtt5_listener_release(m_listener);
+                    m_listener = nullptr;
+                }
+            }
+
+            Mqtt5ListenerCore::~Mqtt5ListenerCore() {}
+
+            Mqtt5ListenerCore::Mqtt5ListenerCore(
+                const Mqtt5ListenerOptions &options,
+                const std::shared_ptr<Mqtt5Client> &client,
+                Allocator *allocator) noexcept
+                : m_callbackFlag(Mqtt5ListenerCore::CallbackFlag::INVOKE), m_listener(nullptr), m_allocator(allocator)
+            {
+
+                if (client == nullptr)
+                {
+                    AWS_LOGF_DEBUG(AWS_LS_MQTT5_GENERAL, "Failed to create Mqtt5ListenerCore. Invalid Client.");
+                    aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+                    return;
+                }
+
+                m_mqtt5Client = client->shared_from_this();
+                if (options.onConnectionFailure)
```

Comment:
Do these need to be conditional assignments?  Copying an untargeted function should work just fine.